Source/WebCore/Modules/cookie-store/CookieStore.cpp

```diff
@@ -211,9 +211,22 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
     Cookie cookie;
     cookie.name = WTFMove(options.name);
     cookie.value = WTFMove(options.value);
-    cookie.domain = options.domain.isNull() ? document.domain() : WTFMove(options.domain);
     cookie.created = WallTime::now().secondsSinceEpoch().milliseconds();
 
+    cookie.domain = options.domain.isNull() ? document.domain() : WTFMove(options.domain);
+    auto& url = document.url();
+    if (!cookie.domain.isNull()) {
+        if (cookie.domain.startsWith('.')) {
+            promise->reject(Exception { TypeError, "The domain must not begin with a '.'"_s });
+            return;
+        }
+        auto host = url.host();
+        if (host != cookie.domain && !host.endsWith(StringView { makeString(".", cookie.domain) })) {
```

Comment:
We can probably do this without allocating another string with a dot prepended.

I don't see handling for this:
If the byte sequence length of encodedDomain is greater than the maximum attribute value size, then return failure.

Also, do we want to check the public suffix list?  If domain is "com" and url.host() is "apple.com" we probably also want to reject the promise.  This may be more of a spec issue though.

Reply:
Good point. I've changed the check to work without allocating a new string and have added handling for the case where the byte sequence length of the domain is greater than the maximum attribute value size. 

About the public suffix list, what you're saying make senses and maybe we should check this. I didn't add it in the newest iteration because it's not in the spec, but perhaps this is something we should discuss.